### PR TITLE
Manually encoded buffer

### PR DESCRIPTION
**Pull request by cullender-skew**
we require the ability to 'manually' encode an ascii buffer from an encoder. The challenge here is that we need to do JIT header encoding before sending a buffer. 

This PR introduces 2 changes to achieve this:

1. on the `Session`, expose a method to just encode a session Header. This allows us to 'manually' encode a buffer outside the session since the header is valid.

2. On the `SessionHeaderEncoder` interface, we expose the ability to just encode the header onto a provided buffer. This allows us to do a JIT session header encoding before we attempt to send an encoded buffer

There is a test case which demonstrates how to use these 2 changes to encode a buffer outside the session and then re-encode the header.

### CODE REVIEW

**Comment by cullender-skew** [comment]
@RichardWarburton is there a different approach you would prefer to take for this?

**Comment by RichardWarburton** [comment]
Sorry @cullender-skew - just busy spinning plates ;) Comments inline.

**Comment by cullender-skew** [comment]
no worries at all @RichardWarburton, thanks for the great work you have been doing with Artio!

File: artio-codecs/src/main/java/uk/co/real_logic/artio/builder/SessionHeaderEncoder.java [hunk 1]
```diff
@@ -148,4 +150,6 @@
 
     boolean hasLastMsgSeqNumProcessed();
 
+    long startMessage(MutableAsciiBuffer buffer, int offset);
```
[the remark this answers is not in the document]
Comment (cullender-skew) [reply artio-codecs/src/main/java/uk/co/real_logic/artio/builder/SessionHeaderEncoder.java]: exactly - the header is encoded with state from the session which may have changed when we attempt to resend a message after backpressure is applied. This also includes `sentSeqNum` which is store in the session

As per the test case
```java
// encode the header of the buffer with the new session state
headerEncoder.startMessage(asciiBuffer, 0);
```

File: artio-core/src/main/java/uk/co/real_logic/artio/session/Session.java [hunk 2]
```diff
@@ -448,20 +448,14 @@ private long logoutAndDisconnect(final DisconnectReason reason)
...
-     * @param encoder the encoder of the message to be sent
-     * @return the position in the stream that corresponds to the end of this message or a negative
-     * number indicating an error status.
-     * @throws IndexOutOfBoundsException if the encoded message is too large, if this happens consider
-     *                                   increasing {@link CommonConfiguration#sessionBufferSize(int)}
+     * @param header the encoder header
+     * @return the encoded sent sequence number for the header
      */
-    public long send(final Encoder encoder)
+    public int encode(final SessionHeaderEncoder header)
```
[the remark this answers is not in the document]
Comment (cullender-skew) [reply artio-core/src/main/java/uk/co/real_logic/artio/session/Session.java]: makes sense, I quite like `prepare`. I will update if no objections

**Comment by RichardWarburton** [comment]
Thanks - Looks good.

**Pull request merged**